From 3c45bafb8607967c05ac9c7f9707a8f15672b7d8 Mon Sep 17 00:00:00 2001
From: Van Do <van.do.xw@renesas.com>
Date: Mon, 29 Oct 2018 11:42:20 +0700
Subject: [PATCH 171/909] thermal: rcar_gen3_thermal: [E3] Update calculation
 formula due to HW evaluation

commit 958bd36e03b70026ac3a33aaa6afc202e7158d2d from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

1/ HW manual changes temperature calculation formula for E3:
 - When CTEMP is less than 24
    T = CTEMP[5:0] * 5.5 - 72
 - When CTEMP is equal to/greater than 24
    T = CTEMP[5:0] * 5 - 60

2/ HW manual changes CTEMP calculation formula for E3:
 - When temperature is less than 60 degree celsius, the formula is:
    CTEMP = ((interrupt temperature) + 72) / 5.5
 - When temperature is equal to/greater than 60 degree celsius,
    the formula is:
    CTEMP = ((interrupt temperature) + 60) / 5

Signed-off-by: Van Do <van.do.xw@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/thermal/rcar_gen3_thermal.c | 29 +++++++++++++++++++++++++++--
 1 file changed, 27 insertions(+), 2 deletions(-)

diff --git a/drivers/thermal/rcar_gen3_thermal.c b/drivers/thermal/rcar_gen3_thermal.c
index 34faae4abeea..ae17e73d770b 100644
--- a/drivers/thermal/rcar_gen3_thermal.c
+++ b/drivers/thermal/rcar_gen3_thermal.c
@@ -266,7 +266,19 @@ static int rcar_gen3_thermal_convert_temp(struct rcar_gen3_thermal_tsc *tsc)
 			}
 			old = new;
 		}
-		mcelsius = MCELSIUS((reg * 5) - 65);
+		/*
+		 * As hardware description, there are 2 formulas to
+		 * calculate temperature on E3 when CTEMP[5:0] is less than
+		 * and greater or equal to 24.
+		 */
+		if (reg < 24)
+			mcelsius = MCELSIUS(((reg * 55) - 720) / 10);
+			/*
+			 * Equal to mcelsius = MCELSIUS((reg * 5.5) - 72)
+			 * to avoid mismatch between float and integer
+			 */
+		else
+			mcelsius = MCELSIUS((reg * 5) - 60);
 	}
 
 	return mcelsius;
@@ -303,7 +315,20 @@ static int rcar_gen3_thermal_mcelsius_to_temp(struct rcar_gen3_thermal_tsc *tsc,
 
 		ctemp = INT_FIXPT(val);
 	} else {
-		ctemp = (celsius + 65) / 5;
+		/*
+		 * Similarly, to calculate register CTEMP[5:0] for E3
+		 * there are 2 formulas to measure CTEMP[5:0] which is
+		 * depending on temperature changes from less than
+		 * to greater or equal to 60 degrees celsius.
+		 */
+		if (celsius < 60)
+			ctemp = (celsius + 72) * 10  / 55;
+			/*
+			 * Equal to ctemp = (celsius + 72) / 5.5
+			 * to avoid mismatch between float and integer
+			 */
+		else
+			ctemp = (celsius + 60) / 5;
 	}
 
 	return ctemp;
-- 
2.17.1

